packages/yarnpkg-fslib/sources/algorithms/copyPromise.ts

```diff
@@ -263,8 +263,13 @@ async function copyFileViaIndex<P1 extends Path, P2 extends Path>(prelayout: Ope
   });
 
   postlayout.push(async () => {
-    if (!indexStat)
+    if (!indexStat) {
       await destinationFs.lutimesPromise(indexPath, defaultTime, defaultTime);
+      const mode = sourceStat.mode & 0o777;
+      if (mode !== 0o644) {
+        await destinationFs.chmodPromise(indexPath, sourceStat.mode);
+      }
```

Comment:
Doesn't that have a risk to corrupt the index if two packages have the same file with different chmod? I think we need to add the chmod to the index file name if it's not 644 to prevent this ([here](https://github.com/yarnpkg/berry/blob/ceed12d9b7e75102fb5884641dbaa055ccc6c064/packages/yarnpkg-fslib/sources/algorithms/copyPromise.ts#L176)).

Reply:
Yes I did not think about that - added one more commit